applications/SwimmingDEMApplication/python_scripts/swimming_DEM_analysis.py

```diff
@@ -609,175 +603,82 @@ def MakeAnalyticsMeasurements(self):
-        coupling_scheme_type = self.pp.CFD_DEM["coupling_scheme_type"].GetString()
-        integration_scheme = self.pp.CFD_DEM["TranslationalIntegrationScheme"].GetString()
-        dem_inlet_option = self.pp.CFD_DEM["dem_inlet_option"].GetBool()
-        interaction_start_time = self.pp.CFD_DEM["interaction_start_time"].GetDouble()
+    def RunMainTemporalLoop(self): # deprecated
+        self.RunSolutionLoop()
 
+    def RunSolutionLoop(self):
         while self.TheSimulationMustGoOn():
+            self.step, self.time = self._GetSolver().AdvanceInTime(self.step, self.time)
+            self.InitializeSolutionStep()
+            self._GetSolver().Predict()
+            self._GetSolver().SolveSolutionStep()
+            self.FinalizeSolutionStep()
+            self.OutputSolutionStep()
 
-            #self.time = self.time + self.Dt
-            self.time = self.fluid_solution._GetSolver().AdvanceInTime(self.time)
-            self.step += 1
-            self.TellTime(self.time)
-
-            if coupling_scheme_type == "UpdatedDEM":
-                time_final_DEM_substepping = self.time + self.Dt
-
-            else:
-                time_final_DEM_substepping = self.time
-
-            #self.PerformEmbeddedOperations() TO-DO: it's crashing
-
-            self.UpdateALEMeshMovement(self.time)
-
-            # solving the fluid part
-            if self.step >= self.GetFirstStepForFluidComputation():
-                self.FluidSolve(
-                    self.time,
-                    solve_system=self.fluid_solve_counter.Tick() and not self.stationarity
-                    )
-
-            # assessing stationarity
-                if self.stationarity_counter.Tick():
-                    self.AssessStationarity()
-
-            # printing if required
-
-            if self.particles_results_counter.Tick():
-                self.io_tools.PrintParticlesResults(
-                    self.pp.variables_to_print_in_file,
-                    self.time,
-                    self.spheres_model_part)
+    def InitializeSolutionStep(self):
+        self.TellTime(self.time)
+        self.PerformInitialDEMStepOperations(self.time)
+        self.disperse_phase_solution.InitializeSolutionStep()
+        if self._GetSolver().CannotIgnoreFluidNow():
+            self.fluid_solution.InitializeSolutionStep()
+        super(SwimmingDEMAnalysis, self).InitializeSolutionStep()
 
-                self.PrintDrag(
-                    self.drag_list,
-                    self.drag_file_output_list,
-                    self.fluid_model_part,
-                    self.time)
+    def FinalizeSolutionStep(self):
+        # printing if required
+        if self._GetSolver().CannotIgnoreFluidNow():
+            self.fluid_solution.FinalizeSolutionStep()
 
-            if self.print_counter_updated_DEM.Tick():
+        self.disperse_phase_solution.FinalizeSolutionStep()
```

Comment:
Perhaps, we could call disparse_phase finalize before the fluid's

Reply:
For the moment, I will keep this ordering. I prefer to modify the fluid (backward coupling) when all operations (FinalizeSolutionStep) have concluded there. In this way I make sure I have the last word.